README.md
![CircleCI](https://img.shields.io/circleci/build/github/Figedi/node-wal2json)
![npm](https://img.shields.io/npm/v/@figedi/node-wal2json)

## node-wal2json

Small nodejs implementation for wal2json. Supports polling-mode ([via pg_logical_slot_get_changes](https://www.postgresql.org/docs/current/logicaldecoding-example.html)) or streaming-mode ([streaming-replication-protocol](https://www.postgresql.org/docs/current/protocol-replication.html)) to read changes from a given replication_slot

### Installation

```
npm i -s @figedi/node-wal2json
```

### Usage

In order to use this library, make sure that you have install all prerequisites:
1. Install [wal2json](https://github.com/eulerto/wal2json), follow the [configuration](https://github.com/eulerto/wal2json#configuration) setup
2. On your postgres-installation, create a user with `REPLICATION`-attribute
3. You are all set, run the library as follows:

#### Polling mode

```typescript
const clientOpts = { replication: "database", connectionString: "postgres://..." };
const wal2json = new PGPollingReplication(clientOpts, { 
  pollTimeoutMs: 100, // poll every 100ms
  slotName: "test_slot", // create slot-name 'test'-slot'
});


// the instance is a readable-stream, so it can be iterated over
for await (const change of wal2json) {
  console.log(change);
}
```

##### Options
The 2nd parameter of PGPollingReplication allows the following options:

| Opt  | Meaning |
| ------------- | ------------- |
| slotName  | The slotname for pg_logical, will be created if not exists  |
| pollTimeoutMs  | Interval for polling for changes in ms  |
| temporary  | (Optional) Creates a temporary slot bound to the connection  |
| destroySlotOnClose  | (Optional) If a permanent slot is chosen and client is stopped, destroy slot  |

-------

**Streaming mode**

```typescript
const clientOpts = { connectionString: "postgres://..." };
const wal2json = new PGStreamingReplication(clientOpts, {
  startLsn: "0/5711970", // start at given LSN
  updateIntervalMs: 1000, // send updates to pg roughly every 1000ms
  slotName: "test_slot", // use slot-name 'test'-slot'
});

// the instance is a readable-stream, so it can be iterated over
for await (const change of wal2json) {
  console.log(change);
}
```

The 2nd parameter of PGStreamingReplication allows the following options:

| Opt  | Meaning |
| ------------- | ------------- |
| slotName  | The slotname for pg_logical, will be created if not exists |
| updateIntervalMs  | (Optional) Interval for sending-back updates to pg, set 0 for reducing updates to the max |
| startLsn  | (Optional) The start-LSN for the replication, note when the slot is already advanced this is ignored |
| autoAckLsn  | (Optional) When set to true, set all received LSN directly as flushed upon next update-interval |
| temporary  | (Optional) Creates a temporary slot bound to the connection  |
| destroySlotOnClose  | (Optional) If a permanent slot is chosen and client is stopped, destroy slot |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.5](https://github.com/Figedi/node-wal2json/compare/v0.0.4...v0.0.5) (2021-11-20)


### Features

* converts legacy polling to readable, removes rxjs ([7e4a204](https://github.com/Figedi/node-wal2json/commit/7e4a204f563241b6811e643f70272b9f3bd89b21))

### [0.0.4](https://github.com/Figedi/node-wal2json/compare/v0.0.3...v0.0.4) (2021-11-18)


### Features

* initial impl for logical replication protocol ([0cb2025](https://github.com/Figedi/node-wal2json/commit/0cb2025fec2ddf776d0b06ef9e7de7e28c6e7531))
* updates docs, finalizes streaming repl ([5ed5c34](https://github.com/Figedi/node-wal2json/commit/5ed5c34e22d5ec6255a3c4bc193482e98e3ac537))


### Bug Fixes

* fixes build errors ([1ff9e68](https://github.com/Figedi/node-wal2json/commit/1ff9e681bc1b07db3b5aa63d1ef6e2391df72a15))
* removes extra transform class ([3801457](https://github.com/Figedi/node-wal2json/commit/3801457447e7bd934f27dddbe30060012d4b783c))

### [0.0.3](https://github.com/Figedi/node-wal2json/compare/v0.0.2...v0.0.3) (2021-11-14)

### [0.0.2](https://github.com/Figedi/node-wal2json/compare/v0.0.1...v0.0.2) (2021-11-14)

### 0.0.1 (2021-11-14)


### Features

* adds dbg, prettier, eslint ([d6538cb](https://github.com/Figedi/node-wal2json/commit/d6538cb1354daf04bb25742d3fa26bc01acb5719))
* impl. cleanup, +adds async-iterators ([81a224f](https://github.com/Figedi/node-wal2json/commit/81a224f04484ff00b3ee9828ab0b8cf161720233))


src/lib/types.ts
import { Duplex, DuplexOptions } from "stream";
import { Submittable, Connection } from "pg";

export interface IPGBaseReplicationOpts {
  slotName: string;
  temporary?: boolean;
  destroySlotOnClose?: boolean;
}

export interface IPGPollingReplicationOpts extends IPGBaseReplicationOpts {
  pollTimeoutMs: number;
}

export interface IPGStreamingReplicationOpts extends IPGBaseReplicationOpts {
  startLsn?: string;
  updateIntervalMs?: number;
  autoAckLsn?: boolean;
}

export interface IChange<T = Record<string, any>> {
  lsn: string;
  timestamp: Date;
  operation: IUpsertOperation<T> | IDeleteOperation<T>;
}

export interface IUpsertOperation<T = Record<string, any>> {
  kind: "update" | "insert";
  schema: string;
  table: string;
  raw: IRawPGLogicalChange;
  parsed: T;
}
export interface IDeleteOperation<T = Record<string, any>> {
  kind: "delete";
  schema: string;
  table: string;
  raw: IRawPGLogicalChange;
  parsed: T;
}

export interface IRawPGLogicalChange {
  kind: "update" | "insert" | "delete";
  schema: string;
  table: string;
  columnnames: string[];
  columntypes: string[];
  columnvalues: string[];
  oldkeys?: {
    keynames: string[];
    keytypes: string[];
    keyvalues: any[];
  };
}

export interface IRawPGLogicalData {
  timestamp: Date;
  change: IRawPGLogicalChange[];
}

export interface IRawPGLogicalRow {
  lsn: string;
  data: string;
}
export interface IRawStreamingRow {
  timestamp: Date;
  lsn: string;
  data: Omit<IRawPGLogicalData, "timestamp">;
}

declare module "pg-copy-streams" {
  export function both(txt: string, options?: DuplexOptions): CopyBothQueryStream;

  export class CopyBothQueryStream extends Duplex implements Submittable {
    submit(connection: Connection): void;
  }
}


src/lib/helpers.ts
import { Client } from "pg";
import {
  IRawStreamingRow,
  IChange,
  IDeleteOperation,
  IRawPGLogicalData,
  IRawPGLogicalRow,
  IUpsertOperation,
} from "./types";

const POSTGRES_EPOCH_2000_01_01_BIGINT = 946684800000n; // milli-seconds <> 1970 -> 2000

export const getNanoseconds = () => {
  return process.hrtime.bigint();
};

export const formatCenturyMicroToDate = (systemClockMicroSeconds: Buffer): Date => {
  const systemClockMillis = systemClockMicroSeconds.readBigInt64BE() / 1000n;

  return new Date(+(systemClockMillis + POSTGRES_EPOCH_2000_01_01_BIGINT).toString());
};

export const nowAsMicroFromEpoch = (): bigint => {
  const systemClockMillis = (BigInt(Date.now()) - POSTGRES_EPOCH_2000_01_01_BIGINT) * 1000n;

  return systemClockMillis;
};

export const initReplicationSlot = async (client: Client, slotName: string, temporary?: boolean) => {
  const results = await client.query("SELECT * FROM pg_replication_slots WHERE slot_name = $1", [slotName]);
  if (!results.rows.length) {
    await client.query("SELECT pg_create_logical_replication_slot($1, 'wal2json', $2)", [slotName, temporary ?? false]);
  }
};

export const destroyReplicationSlot = async (client: Client, slotName: string) => {
  await client.query("SELECT pg_drop_replication_slot($1)", [slotName]);
};

const convertRawPgLogicalData = <T>(lsn: string, data: IRawPGLogicalData) => {
  return data.change.map(change => {
    let operation: IDeleteOperation<T> | IUpsertOperation<T>;
    if (change.kind === "delete") {
      operation = {
        kind: change.kind,
        schema: change.schema,
        table: change.table,
        raw: change,
        parsed: change.oldkeys!.keynames.reduce(
          (acc: any, name: string, i: number) => ({
            ...acc,
            [name]: change.oldkeys!.keyvalues[i],
          }),
          {} as T,
        ),
      };
    } else {
      operation = {
        kind: change.kind,
        schema: change.schema,
        table: change.table,
        raw: change,
        parsed: change.columnnames.reduce(
          (acc: any, name: string, i: number) => ({
            ...acc,
            [name]: change.columnvalues[i],
          }),
          {} as T,
        ),
      };
    }
    return {
      lsn,
      timestamp: new Date(data.timestamp),
      operation,
    };
  });
};

export const normalizeStreamResults = <T>(row: IRawStreamingRow): IChange<T>[] => {
  return convertRawPgLogicalData(row.lsn, { ...row.data, timestamp: row.timestamp });
};

export const normalizePollingResults = <T>(rows: IRawPGLogicalRow[]): IChange<T>[] => {
  return rows.flatMap(row => {
    const data = JSON.parse(row.data) as IRawPGLogicalData;

    return convertRawPgLogicalData(row.lsn, data);
  });
};


src/lib/Wal2JSON.ts
import { Client, ClientConfig } from "pg";
import { EMPTY, from, Observable } from "rxjs";
import _debug from "debug";

const debug = _debug("wal2json");

import {
  IRawPGLogicalData,
  IRawPGLogicalRow,
  IChange,
  IDeleteOperation,
  IUpsertOperation,
  IWal2JSONOpts,
} from "./types";

export class Wal2JSON<T extends Record<string, any> = Record<string, any>> {
  private running = false;

  private client!: Client;

  constructor(clientOpts: string | ClientConfig | Client, private opts: IWal2JSONOpts) {
    if (clientOpts instanceof Client) {
      this.client = clientOpts;
    } else {
      this.client = new Client(clientOpts);
    }
    if (this.opts.pollTimeoutMs < 100) {
      debug(
        `Very short timeout (${this.opts.pollTimeoutMs}ms) chosen, this can lead to ` +
          "overwhelming your postgres instance",
      );
    }
  }

  public asAsyncIterator(): AsyncIterable<IChange<T>> {
    if (this.running) {
      void this.onError(
        new Error("Service is already running, please call stop() first if you wish to consume via Async-iterators"),
      );
      return {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        async *[Symbol.asyncIterator]() {}, // empty async-iteratable, returns right away
      };
    }
    this.running = true;
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const self = this;
    return {
      async *[Symbol.asyncIterator]() {
        await self.start();

        while (self.running) {
          const changes = await self.readChanges();
          yield* changes;

          await new Promise(resolve => setTimeout(resolve, self.opts.pollTimeoutMs));
        }
      },
    };
  }

  public asObservable(): Observable<IChange<T>> {
    if (this.running) {
      void this.onError(
        new Error("Service is already running, please call stop() first if you wish to consume via Observables"),
      );
      return EMPTY;
    }

    return from(this.asAsyncIterator());
  }

  private async initReplicationSlot() {
    const results = await this.client.query("SELECT * FROM pg_replication_slots WHERE slot_name = $1", [
      this.opts.slotName,
    ]);
    if (!results.rows.length) {
      await this.client.query("SELECT pg_create_logical_replication_slot($1, 'wal2json', $2)", [
        this.opts.slotName,
        this.opts.temporary ?? false,
      ]);
    }
  }

  private async destroyReplicationSlot() {
    await this.client.query("SELECT pg_drop_replication_slot($1)", [this.opts.slotName]);
  }

  private normalizeResults(rows: IRawPGLogicalRow[]): IChange<T>[] {
    return rows.flatMap(row => {
      const data = JSON.parse(row.data) as IRawPGLogicalData;

      return data.change.map(change => {
        let operation: IDeleteOperation<T> | IUpsertOperation<T>;
        if (change.kind === "delete") {
          operation = {
            kind: change.kind,
            schema: change.schema,
            table: change.table,
            raw: change,
            parsed: change.oldkeys!.keynames.reduce(
              (acc: any, name: string, i: number) => ({
                ...acc,
                [name]: change.oldkeys!.keyvalues[i],
              }),
              {} as T,
            ),
          };
        } else {
          operation = {
            kind: change.kind,
            schema: change.schema,
            table: change.table,
            raw: change,
            parsed: change.columnnames.reduce(
              (acc: any, name: string, i: number) => ({
                ...acc,
                [name]: change.columnvalues[i],
              }),
              {} as T,
            ),
          };
        }
        return {
          lsn: row.lsn,
          xid: row.xid,
          timestamp: new Date(data.timestamp),
          operation,
        };
      });
    });
  }

  private readChanges = async (): Promise<IChange<T>[]> => {
    // @todo do not use internal apis from pg anymore
    // eslint-disable-next-line no-underscore-dangle
    if (!(this.client as any).readyForQuery || (this.client as any)._ending) {
      return Promise.resolve([]);
    }
    try {
      const results = await this.client.query(
        "SELECT * FROM pg_logical_slot_get_changes($1, NULL, NULL, 'include-timestamp', '1')",
        [this.opts.slotName],
      );
      const normalizedRows = this.normalizeResults(results.rows);
      if (normalizedRows.length) {
        debug({ normalizedRows }, `Received ${normalizedRows.length} changed rows from postgres`);
      }
      return normalizedRows;
    } catch (e) {
      await this.onError(e);
      throw e;
    }
  };

  private async onError(e: Error) {
    debug(e);
    await this.close();
    throw e;
  }

  private async close() {
    debug("Gracefully closing service");
    this.running = false;
    if (this.opts.destroySlotOnClose) {
      await this.destroyReplicationSlot();
    }
    await this.client.end();
    debug("Service successfully closed");
  }

  private async start(): Promise<void> {
    debug(this.opts, `Trying to initialize service for replication_slot ${this.opts.slotName}`);
    try {
      await this.client.connect();
      await this.initReplicationSlot();
      debug(`Service successfully initialized with replication_slot ${this.opts.slotName}`);
    } catch (e: any) {
      return this.onError(e);
    }
  }

  public async stop() {
    if (!this.running) {
      return this.onError(new Error("Service is not running yet, pleaase call start() firsts"));
    }
    return this.close();
  }
}


src/lib/PGLogical.ts
import { Client } from "pg";
import { Transform, pipeline, finished } from "stream";
import { both, CopyBothQueryStream } from "pg-copy-streams";
import _debug from "debug";

const debug = _debug("pglogical");
// eslint-disable-next-line no-underscore-dangle
const BufferList = require("obuf");

const POSTGRES_EPOCH_2000_01_01_BIGINT = 946684800000n; // milli-seconds <> 1970 -> 2000

const getNanoseconds = () => {
  return process.hrtime.bigint();
};

export class LogSequenceNumber {
  private constructor(public buf: Buffer) {}

  public static INVALID_LSN = LogSequenceNumber.fromBuf(Buffer.alloc(0));

  public get empty(): boolean {
    return !this.buf.length || this.buf.readInt8() === 0;
  }

  public static fromBuf(buf: Buffer): LogSequenceNumber {
    return new LogSequenceNumber(buf);
  }

  public static fromString(lsn: string): LogSequenceNumber {
    const [logicalXLogStr, segmentStr] = lsn.split("/");

    if (!logicalXLogStr || !segmentStr) {
      return LogSequenceNumber.INVALID_LSN;
    }

    const logicalXlog = parseInt(logicalXLogStr, 16);
    const segment = parseInt(segmentStr, 16);
    const buf = Buffer.alloc(8); // 8 bytes -> 1 int64
    buf.writeInt32BE(logicalXlog); // first 4 bytes
    buf.writeInt32BE(segment, 4); // next 4 bytes

    return new LogSequenceNumber(buf);
  }

  public toString() {
    if (!this.buf.length) {
      return "0/0";
    }

    const logicalXlog = this.buf.readInt32BE(0);
    const segment = this.buf.readInt32BE(4);
    return `${logicalXlog.toString(16)}/${segment.toString(16)}`;
  }

  public asInt64() {
    if (!this.buf?.length) {
      return 0n;
    }
    return this.buf.readBigInt64BE();
  }
}

const formatCenturyMicroToDate = (systemClockMicroSeconds: Buffer): Date => {
  const systemClockMillis = systemClockMicroSeconds.readBigInt64BE() / 1000n;

  return new Date(+(systemClockMillis + POSTGRES_EPOCH_2000_01_01_BIGINT).toString());
};

const nowAsMicroFromEpoch = (): bigint => {
  const systemClockMillis = (BigInt(Date.now()) - POSTGRES_EPOCH_2000_01_01_BIGINT) * 1000n;

  return systemClockMillis;
};

const PG_CODE = 1;
const PG_MESSAGE = 2;
class PgLogicalParser extends Transform {
  private buf = new BufferList();

  private state = PG_CODE;

  private code: number | null = null;

  private lastServerLSN: LogSequenceNumber = LogSequenceNumber.INVALID_LSN;

  private lastReceiveLSN: LogSequenceNumber = LogSequenceNumber.INVALID_LSN;

  public _flush() {}

  constructor(private cb: (data: any) => void) {
    super();
  }

  public _transform(chunk: Buffer, _encoding: string, callback: (error?: Error) => void) {
    this.buf.push(chunk);
    while (this.buf.size > 0) {
      if (this.state === PG_CODE) {
        if (!this.buf.has(1)) break;
        this.code = this.buf.readUInt8();
        this.state = PG_MESSAGE;
      }
      if (this.state === PG_MESSAGE) {
        if (this.code === 0x6b /*k*/) {
          this.lastServerLSN = LogSequenceNumber.fromBuf(this.buf.take(8));
          if (this.lastServerLSN.asInt64() > this.lastReceiveLSN.asInt64()) {
            this.lastReceiveLSN = this.lastServerLSN;
          }
          const systemClock: Buffer = this.buf.take(8);

          const replyRequired = this.buf.take(1).readUInt8() !== 0;

          this.cb({
            lastReceiveLSN: this.lastReceiveLSN,
            systemClock: formatCenturyMicroToDate(systemClock),
            replyRequired, // @todo either when this is true or when a configured timeout (based on the lastStatusUpdate diff -> update)
          });
          this.state = PG_CODE;
          // x
        } else if (this.code === 0x77 /*w*/) {
          this.lastReceiveLSN = LogSequenceNumber.fromBuf(this.buf.take(8));
          this.lastServerLSN = LogSequenceNumber.fromBuf(this.buf.take(8));
          const systemClock: Buffer = this.buf.take(8);

          const data = this.buf.take(this.buf.size); /* plugin data */

          this.cb({
            lastReceiveLSN: this.lastReceiveLSN,
            systemClock: formatCenturyMicroToDate(systemClock),
            data: JSON.parse(data.toString("utf-8")),
          });
          this.state = PG_CODE;
        } else {
          return callback(new Error("wrong message code inside"));
        }
      }
      break;
    }
    callback();
  }
}

export class PgLogical {
  private parser = new PgLogicalParser(data => this.onDataReceived(data));

  private copyBothStream!: CopyBothQueryStream;

  private lastReceiveLSN = LogSequenceNumber.INVALID_LSN;

  private lastFlushedLSN = LogSequenceNumber.INVALID_LSN;

  private lastAppliedLSN = LogSequenceNumber.INVALID_LSN;

  public lastStatusUpdate?: bigint;

  private interval?: NodeJS.Timer;

  constructor(private client: Client, private slotName: string, private startLsn: string) {
    this.lastReceiveLSN = LogSequenceNumber.fromString(startLsn);

    this.interval = setInterval(() => this.updateStatus(), 1000);
  }

  public start() {
    this.copyBothStream = both(`START_REPLICATION SLOT ${this.slotName} LOGICAL ${this.startLsn}`, {
      alignOnCopyDataFrame: true,
    } as any);
    this.client.query(this.copyBothStream);

    pipeline(this.copyBothStream, this.parser, error => {
      if (error) {
        debug({ error }, "Unexpected error in parser-stream");
      } else {
        debug("Parser-stream ended");
      }
    });
    finished(this.copyBothStream, error => {
      if (error) {
        debug({ error }, "Unexpected error in copy-stream");
      } else {
        debug("Copy-stream ended");
      }
      this.client.end();
    });
  }

  public stop() {
    if (this.interval) {
      clearInterval(this.interval);
    }
    this.copyBothStream.end();
  }

  public updateStatus() {
    if (!this.lastReceiveLSN.empty) {
      return;
    }
    const buf = Buffer.alloc(1 + 8 + 8 + 8 + 8 + 1);
    buf.write("r"); // 0x72, indicates Standby status update
    buf.writeBigInt64BE(this.lastReceiveLSN.asInt64(), 1);
    buf.writeBigInt64BE(this.lastFlushedLSN.asInt64(), 9);
    buf.writeBigInt64BE(this.lastAppliedLSN.asInt64(), 17);
    buf.writeBigInt64BE(nowAsMicroFromEpoch(), 25);
    buf.writeInt8(1, 33); // no reply required

    this.copyBothStream.write(buf);

    this.lastStatusUpdate = getNanoseconds();
  }

  private onDataReceived({ lastReceiveLSN, systemClock, data }: any) {
    this.lastReceiveLSN = lastReceiveLSN;
    this.lastFlushedLSN = lastReceiveLSN;
    this.lastAppliedLSN = lastReceiveLSN;

    if (data) {
      console.log("onDataReceived", {
        timestamp: systemClock,
        lsn: lastReceiveLSN.toString(),
        data,
      });
    } else {
      // console.log("onHeartbeatReceived", { timestamp: systemClock, lsn: lastReceiveLSN.toString() });
    }
  }
}


src/lib/LogSequenceNumber.ts
export class LogSequenceNumber {
  private constructor(public buffer: Buffer) {}

  public static INVALID_LSN = LogSequenceNumber.fromBuf(Buffer.alloc(0));

  public get empty(): boolean {
    return !this.buffer.length || this.buffer.readInt8() === 0;
  }

  public static fromBuf(buf: Buffer): LogSequenceNumber {
    return new LogSequenceNumber(buf);
  }

  public static fromString(lsn: string): LogSequenceNumber {
    const [logicalXLogStr, segmentStr] = lsn.split("/");

    if (!logicalXLogStr || !segmentStr) {
      return LogSequenceNumber.INVALID_LSN;
    }

    const logicalXlog = parseInt(logicalXLogStr, 16);
    const segment = parseInt(segmentStr, 16);
    const buf = Buffer.alloc(8); // 8 bytes -> 1 int64
    buf.writeInt32BE(logicalXlog); // first 4 bytes
    buf.writeInt32BE(segment, 4); // next 4 bytes

    return new LogSequenceNumber(buf);
  }

  public toString() {
    if (!this.buffer.length) {
      return "0/0";
    }

    const logicalXlog = this.buffer.readInt32BE(0);
    const segment = this.buffer.readInt32BE(4);
    return `${logicalXlog.toString(16)}/${segment.toString(16)}`;
  }

  public asInt64() {
    if (!this.buffer?.length) {
      return 0n;
    }
    return this.buffer.readBigInt64BE();
  }
}


src/lib/PGPollingReplication.ts
import { Client, ClientConfig } from "pg";
import { Readable } from "stream";
import _debug from "debug";

const debug = _debug("wal2json");

import { IChange, IPGPollingReplicationOpts } from "./types";
import { destroyReplicationSlot, initReplicationSlot, normalizePollingResults } from "./helpers";

export class PGPollingReplication<T extends Record<string, any> = Record<string, any>> extends Readable {
  private running = false;

  private client!: Client;

  constructor(clientOpts: string | ClientConfig, private opts: IPGPollingReplicationOpts) {
    super({ objectMode: true });

    this.client = new Client(clientOpts);
    if (this.opts.pollTimeoutMs < 100) {
      debug(
        `Very short timeout (${this.opts.pollTimeoutMs}ms) chosen, this can lead to ` +
          "overwhelming your postgres instance",
      );
    }
  }

  public async start(): Promise<void> {
    debug(this.opts, `Trying to initialize service for replication_slot ${this.opts.slotName}`);
    try {
      await this.client.connect();
      await initReplicationSlot(this.client, this.opts.slotName, this.opts.temporary);
      this.running = true;
      debug(`Service successfully initialized with replication_slot ${this.opts.slotName}`);
      this.initPolling().catch(e => this.onError(e));
    } catch (e: any) {
      return this.onError(e);
    }
  }

  public _read() {}

  public async stop() {
    if (!this.running) {
      return this.onError(new Error("Service is not running yet, pleaase call start() firsts"));
    }
    return this.close();
  }

  private async initPolling() {
    while (this.running) {
      const changes = await this.readChanges();
      changes.forEach(change => this.push(change));

      await new Promise(resolve => setTimeout(resolve, this.opts.pollTimeoutMs));
    }
  }

  private readChanges = async (): Promise<IChange<T>[]> => {
    // @todo do not use internal apis from pg anymore
    // eslint-disable-next-line no-underscore-dangle
    if (!(this.client as any).readyForQuery || (this.client as any)._ending) {
      return Promise.resolve([]);
    }
    try {
      const results = await this.client.query(
        "SELECT * FROM pg_logical_slot_get_changes($1, NULL, NULL, 'include-timestamp', '1')",
        [this.opts.slotName],
      );
      const normalizedRows = normalizePollingResults<T>(results.rows);
      if (normalizedRows.length) {
        debug({ normalizedRows }, `Received ${normalizedRows.length} changed rows from postgres`);
      }
      return normalizedRows;
    } catch (e) {
      await this.onError(e);
      throw e;
    }
  };

  private async onError(e: Error) {
    debug(e);
    await this.close();
    throw e;
  }

  private async close() {
    debug("Gracefully closing service");
    this.running = false;
    if (this.opts.destroySlotOnClose) {
      await destroyReplicationSlot(this.client, this.opts.slotName);
    }
    await this.client.end();
    debug("Service successfully closed");
  }
}


src/lib/PGStreamingReplication.ts
import { Client, ClientConfig } from "pg";
import { Transform, pipeline, finished } from "stream";
import { both, CopyBothQueryStream } from "pg-copy-streams";
import _debug from "debug";
import assert from "assert";

import {
  destroyReplicationSlot,
  formatCenturyMicroToDate,
  getNanoseconds,
  initReplicationSlot,
  normalizeStreamResults,
  nowAsMicroFromEpoch,
} from "./helpers";
import { IPGStreamingReplicationOpts } from "./types";
import { LogSequenceNumber } from "./LogSequenceNumber";

const debug = _debug("pglogical");
const BufferList = require("obuf");

const PG_CODE = 1;
const PG_MESSAGE = 2;

export class PGStreamingReplication<T extends Record<string, any> = Record<string, any>> extends Transform {
  private started = false;

  private buf!: any;

  private state = PG_CODE;

  private code: number | null = null;

  private copyBothStream!: CopyBothQueryStream;

  private lastStatusUpdate!: bigint;

  private updateIntervalMs!: number;

  private lastServerLSN = LogSequenceNumber.INVALID_LSN;

  private lastReceivedLSN = LogSequenceNumber.INVALID_LSN;

  private lastFlushedLSN = LogSequenceNumber.INVALID_LSN;

  private lastAppliedLSN = LogSequenceNumber.INVALID_LSN;

  private client!: Client;

  constructor(clientOpts: string | ClientConfig, private opts: IPGStreamingReplicationOpts) {
    super({ objectMode: true });

    this.client = new Client(clientOpts);
    assert(
      (this.client as any).connectionParameters.replication === "database",
      "Need to pass replication: 'database' to pg-client",
    );

    this.buf = new BufferList();
    this.updateIntervalMs = this.opts.updateIntervalMs ?? 0;
    this.lastReceivedLSN = this.opts.startLsn
      ? LogSequenceNumber.fromString(this.opts.startLsn)
      : LogSequenceNumber.INVALID_LSN;
    this.lastStatusUpdate = getNanoseconds() - BigInt(this.updateIntervalMs) * 1000000n;
  }

  public _flush() {}

  /**
   * Main-hook for receiving chunks via the upstream copy-stream. See sub-package handlers
   * for the behaviour. Mostly copied from pg-copy-streams example
   */
  public _transform(chunk: Buffer, _encoding: string, next: (error?: Error) => void) {
    this.buf.push(chunk);
    while (this.buf.size > 0) {
      if (this.state === PG_CODE) {
        if (!this.buf.has(1)) break;
        this.code = this.buf.readUInt8();
        this.state = PG_MESSAGE;
      }
      if (this.state === PG_MESSAGE) {
        if (this.code === 0x6b /*k*/) {
          this.onKeepAlivePacketReceived();
          // x
        } else if (this.code === 0x77 /*w*/) {
          this.onXLogDataPacketReceived();
        } else {
          return next(new Error("wrong message code inside"));
        }
      }

      break;
    }
    next();
  }

  /**
   * Starts consumption of replication-logs
   */
  public async start() {
    if (this.started) {
      throw new Error("Cannot call PGStreamingReplication#start() twice, please call stop() first");
    }
    await this.client.connect();

    const startLsn = await this.initSlotWithStartLSN();
    this.copyBothStream = both(`START_REPLICATION SLOT ${this.opts.slotName} LOGICAL ${startLsn}`, {
      alignOnCopyDataFrame: true,
    } as any);
    this.client.query(this.copyBothStream);

    pipeline(this.copyBothStream, this, error => {
      if (error) {
        debug({ error }, "Unexpected error in parser-stream");
      } else {
        debug("Parser-stream ended");
      }
    });
    finished(this.copyBothStream, error => {
      if (error) {
        debug({ error }, "Unexpected error in copy-stream");
      } else {
        debug("Copy-stream ended");
      }

      this.client.end();
      return this.cleanup();
    });

    this.started = true;
  }

  /**
   * Stops replication-log consumption by stopping the upstream copy-stream
   */
  public stop() {
    this.copyBothStream.end();
  }

  /**
   * Set flushed LSN.
   * This parameter will be sent to backend on next update status iteration.
   * Flushed LSN position help backend define which WAL can be recycled.
   */
  public setLastFlushedLSN(lsn: LogSequenceNumber) {
    this.lastFlushedLSN = lsn;
  }

  /**
   * Set last-applied LSN
   * Inform backend which LSN has been applied on standby.
   * Feedback will send to backend on next update status iteration.
   */
  public setLastAppliedLSN(lsn: LogSequenceNumber) {
    this.lastAppliedLSN = lsn;
  }

  private getQueryClient(): Client {
    return new Client({
      user: this.client.user,
      password: this.client.password,
      host: this.client.host,
      port: this.client.port,
      database: this.client.database,
    });
  }

  private async initSlotWithStartLSN(): Promise<string> {
    const client = this.getQueryClient();
    try {
      await client.connect();
      await initReplicationSlot(client, this.opts.slotName);

      if (this.opts.startLsn) {
        return this.opts.startLsn;
      }

      const {
        rows: [row],
      } = await client.query("SELECT pg_current_wal_lsn()");
      if (!row.pg_current_wal_lsn) {
        throw new Error(`Did not find last pg_current_wal_lsn for slot-name ${this.opts.slotName}`);
      }
      this.lastReceivedLSN = LogSequenceNumber.fromString(row.pg_current_wal_lsn);
      return row.pg_current_wal_lsn;
    } finally {
      await client.end();
    }
  }

  /**
   * Processes an xlog-data packet from pg by normalizing the logical replication
   * data as json
   */
  private onXLogDataPacketReceived() {
    this.lastReceivedLSN = LogSequenceNumber.fromBuf(this.buf.take(8));
    this.lastServerLSN = LogSequenceNumber.fromBuf(this.buf.take(8));
    const systemClock: Buffer = this.buf.take(8);
    const data = this.buf.take(this.buf.size);

    debug({ systemClock, lastReceiveLSN: this.lastReceivedLSN }, "Received XLogData message");

    if (this.updateDue()) {
      this.updateStatus();
    }

    const normalizedRow = normalizeStreamResults<T>({
      timestamp: formatCenturyMicroToDate(systemClock),
      lsn: this.lastReceivedLSN.toString(),
      data: JSON.parse(data.toString("utf-8")),
    });
    this.push(normalizedRow);

    this.state = PG_CODE;
  }

  /**
   * Processes a keep-alive packet from pg. If pg indicates a status update is
   * required or the set up interval is due, this method sends back the currently
   * consumed LSN's
   */
  private onKeepAlivePacketReceived() {
    this.lastServerLSN = LogSequenceNumber.fromBuf(this.buf.take(8));
    const systemClock: Buffer = this.buf.take(8);
    const replyRequired = this.buf.take(1).readUInt8() !== 0;

    if (this.lastServerLSN.asInt64() > this.lastReceivedLSN.asInt64()) {
      this.lastReceivedLSN = this.lastServerLSN;
    }

    debug({ systemClock, replyRequired, lastReceiveLSN: this.lastReceivedLSN }, "Received Primary keepalive message");

    if (replyRequired || this.updateDue()) {
      this.updateStatus();
    }

    this.state = PG_CODE;
  }

  /**
   * Determines whether an update is due based on the last
   * update and the configured status-updates. If updateIntervalMs
   * is set to 0, no update will be performed (unless required from pg)
   */
  private updateDue(): boolean {
    if (this.updateIntervalMs == 0) {
      return false;
    }
    const diffMs = (getNanoseconds() - this.lastStatusUpdate) / BigInt(1e6);
    return diffMs >= this.updateIntervalMs;
  }

  /**
   * Sends back to pg a 'Standby status update'. This mainly
   * acknowledges a received-lsn by the stream
   */
  private updateStatus() {
    if (!this.lastReceivedLSN.empty) {
      return;
    }
    if (this.opts.autoAckLsn) {
      this.lastFlushedLSN = this.lastReceivedLSN;
      this.lastAppliedLSN = this.lastReceivedLSN;
    }
    const buf = Buffer.alloc(1 + 8 + 8 + 8 + 8 + 1);
    buf.write("r"); // 0x72, indicates Standby status update
    buf.writeBigInt64BE(this.lastReceivedLSN.asInt64(), 1);
    buf.writeBigInt64BE(this.lastFlushedLSN.asInt64(), 9);
    buf.writeBigInt64BE(this.lastAppliedLSN.asInt64(), 17);
    buf.writeBigInt64BE(nowAsMicroFromEpoch(), 25);
    buf.writeInt8(1, 33); // no reply required

    this.copyBothStream.write(buf);

    this.lastStatusUpdate = getNanoseconds();
  }

  private async cleanup() {
    if (this.opts.destroySlotOnClose) {
      const client = this.getQueryClient();
      try {
        await client.connect();
        await destroyReplicationSlot(client, this.opts.slotName);
      } finally {
        await client.end();
      }
    }
  }
}
